### PR TITLE
Exploit chain 1

### PR DESCRIPTION
**Pull request by shoebpate1**
This needs improvement regarding verification of challenge is solved or not and one client side unit test.
Please review @J12934 @wurstbrot @bkimminich :)

### CODE REVIEW

**Comment by J12934** [comment]
Hi Shoeb
will take a closer look at this later today.
But at first glance i noticed another security flaw in that code: 😉

An attacker can avoid the logging of his IP by never logging out.😅
It would be nice if we could make a challenge out of that (or a other way to prevent the ip logging), but that's also a pretty hard thing to verify whether it was solved or not

**Comment by shoebpate1** [comment]
Great, By the way I did not understand about security flaw of not logging out.

**Comment by J12934** [comment]
Well the IP just gets logged when the user clicks the logout button.
So, the IP logging as a protection / logging mechanism can be avoided by just not clicking the button.

**Comment by bkimminich** [comment]
How about we put the last IP of a user on the administration page user list? That way it could be verified. Name of that challenge should involve the word "ninja"... 🤣

File: app/js/services/LogoutService.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+angular.module('juiceShop').factory('LogoutService', ['$http', '$q', function ($http, $q) {
+  'use strict'
+
+  var host = '/rest/saveLoginIp'
```
Comment (wurstbrot) [inline app/js/services/LogoutService.js:4]: Why is this variable named host?

File: models/user.js [hunk 2]
```diff
@@ -24,6 +24,10 @@ module.exports = (sequelize, {STRING, BOOLEAN}) => {
     isAdmin: {
       type: BOOLEAN,
       defaultValue: false
+    },
+    lastLoginIp: {
+      type: STRING,
+      defaultValue: 'x.x.x.x'
```
Comment (wurstbrot) [inline models/user.js]: I would prefer a number like 0.0.0.0 or 127.0.0.1

File: test/e2e/restApiSpec.js [hunk 3]
```diff
@@ -36,6 +36,19 @@ describe('/rest', () => {
     protractor.expect.challengeSolved({challenge: 'XSS Tier 3'})
   })
 
+  describe('challenge "xss5"', () => {
+    protractor.beforeEach.login({email: 'admin@' + config.get('application.domain'), password: 'admin123'})
+
+    it('should be possible to save log-in IP when logged in', () => {
```
Comment (wurstbrot) [inline test/e2e/restApiSpec.js:42]: I wish a test like 'it should not be possible to save log-in IP when not logged in'

**Review by wurstbrot — CHANGES_REQUESTED** [review]
I have the following suggestions

**Comment by shoebpate1** [comment]
Will get on it @wurstbrot , Any suggestion on how to pass the failing client side test ? @bkimminich @J12934

File: app/js/services/LogoutService.js (hunk 1 above)
Comment (shoebpate1) [reply app/js/services/LogoutService.js:4]: host variable is used for base url in every service.

File: test/e2e/restApiSpec.js (hunk 3 above)
Comment (shoebpate1) [reply test/e2e/restApiSpec.js:42]: Wrote it :)

**Comment by bkimminich** [comment]
@wurstbrot / @J12934, is this one ready from your point of view so we can merge it?

**Comment by J12934** [comment]
I don’t think it is ready yet.
We are still unsure about how to properly validate whether the challenge was solved or not. The current way only checks part of the challenge.

**Pull request merged**